### PR TITLE
[camerax] Implement onCameraClosing

### PR DESCRIPTION
**Pull request by camsim99**
### Overview

Continuation of https://github.com/flutter/packages/pull/3419.

Implements onCameraClosing that will send an event anytime the camera closes, as detected by observing the LiveData of the CameraState of the current camera being used in the plugin.

Fixes https://github.com/flutter/flutter/issues/121162.

### Highlights of what has changed since the latest reviews of the [last PR](https://github.com/flutter/packages/pull/3419):

- Wrapped `LiveData#getValue` ([docs](https://developer.android.com/reference/androidx/lifecycle/LiveData#getValue())) for the purposes of retrieving current value `LiveData<ZoomState>` to properly implement retrieving zoom information
- Modified Dart Host API and FlutterAPI implementations' documentation on their `binaryMessenger` and `instanceManager` documentation to be clearer as per suggestion from comments on the original PR
- Changed `binaryMessenger` and `instanceManager` fields to be private in Flutter API implementations
- Removed `cast` method and added enum `LiveDataSupportedType` to pigeon file to use for safely casting the `LiveData` generic when received on the Dart or native side.
- Moved `CameraStateError` code decoding to Dart side as this was written by me and not directly provided by CameraX.
- Added tests for the conversion of objects.
- The following issues were filed to capture feedback given but not addressed in this PR: https://github.com/flutter/flutter/issues/125926, https://github.com/flutter/flutter/issues/125928.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/camera/camera_android_camerax/test/live_data_test.dart
Comment (camsim99) [inline packages/camera/camera_android_camerax/test/live_data_test.dart:106]: @bparrishMines I ended up needing to set this up here because setting it at the top of the main method didn't work, but it also didn't work without it at the top of the file. Do you know why or is there some technical details I'm missing?

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/CameraStateTest.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/CameraStateTest.java:97]: Love this. You might consider including the type in fail message so it is easy to see what value was failing .

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/CameraStateTest.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/CameraStateTest.java:76]: No action needed: 
Another patten you might see in code would reduce the duplicate calls to  `assertEquals(CameraStateFlutterApiWrapper.getCameraStateType(type), CameraStateType.<THING>);`. It can be easier to read if you break the conditions into a map (or pair) of {CameraState.Type to expected CameraStateType} that validates the known conversions. Then another test that loops over CameraState.Type.values() and verifies that all the values are non null and do not throw an exception.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/LiveDataHostApiImpl.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/LiveDataHostApiImpl.java:75]: Non blocking for this pr. I think what we want to do here is make all supported types declare a string type that we can switch on then let callers register a type handler. See [command pattern](https://en.wikipedia.org/wiki/Command_pattern) then the live data instance delegates to the registered handlers so that each type can handle its own data. Potentially even in its own class as an extension function. 

For this pr can I mildly prefer extracting the logic for each sub type conversion into their own methods.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/LiveDataHostApiImpl.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/LiveDataHostApiImpl.java:1]: I really like this modified version without unchecked casting.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ObserverFlutterApiWrapper.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ObserverFlutterApiWrapper.java:58]: It looks like the types need to be supported here an in `plugins/camerax/LiveDataHostApiImpl.java`. Can you define a list or some other object that is shared in the java files for what types are supported? If not can you add comments that link the implementation requirements for adding a new type?

Comment (reidbaker) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ObserverFlutterApiWrapper.java:58]: Also I would suggest refactoring these conversions into their own methods used in this class.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ObserverFlutterApiWrapper.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ObserverFlutterApiWrapper.java:54]: Can this take in a `Object<T>`value?

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/ObserverTest.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/ObserverTest.java:50]: Do these still need to be here?

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/ImageCaptureTest.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/ImageCaptureTest.java:193]: Why did these lines need to be removed for this PR? Unrelated test failure?

**Review by reidbaker — COMMENTED** [review]
java code reviewed.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:115]: You left a comment about this in the old PR :)

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/CameraStateTest.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/CameraStateTest.java:76]: Added to https://github.com/flutter/flutter/issues/125928

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/ImageCaptureTest.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/ImageCaptureTest.java:193]: A warning indicated these values aren't used. Unrelated fix!

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (bparrishMines) [inline packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:568]: This should use a weak reference to the AndroidCameraCameraX instance. Something like:

```dart
final WeakReference<AndroidCameraCameraX> weakThis = WeakReference<AndroidCameraCameraX>(this);
```

The reason being that callback methods in the Native Wrapper API can cause leaking if they end up referencing themselves. Which there is a good chance to happen if you reference the complex class `AndroidCameraCameraX`.

I forgot to mention in the image streaming PR, so the code for that may need to be updated as well.

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/ObserverTest.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/ObserverTest.java:50]: Nope! :)

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ObserverFlutterApiWrapper.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ObserverFlutterApiWrapper.java:63]: I added a comment pointing to the instructions I wrote in `pigeons/camerax_library.dart`. I can't use the enum directly here because this is a callback from Android, but it is defined in `GeneratedCameraXLibrary.java`. I also mentioned it in the comment.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ObserverFlutterApiWrapper.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ObserverFlutterApiWrapper.java]: I don't think that's legal, but I changed the types to reference use `T`

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/LiveDataHostApiImpl.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/LiveDataHostApiImpl.java:75]: I extracted it out the logic for now, but I definitely think we should change this as per your suggestion, especially since this sort of logic is shared across classes. Added to https://github.com/flutter/flutter/issues/125926

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:570]: Updated both!

Comment (reidbaker) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:570]: Not blocking this pr but if it is important to use a weak reference to make sure something does not leak then we should add a test that covers that behavior.

Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:570]: To clarify, test that a weak reference is used in these callbacks and/or that using a weak reference avoids any leaking?

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (bparrishMines) [inline packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:335]: Would it not be better to have a local `zoomState` variable that is updated by an `Observer` added to `liveZoomState`? Same for `getMinZoomLevel`.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (bparrishMines) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:570]: Garbage collection for Dart is non-deterministic, so testing for it a little difficult. I do have a test for leaking in the `webview_flutter` integration tests that verifies the `WebView` is garbage collected: https://github.com/flutter/packages/blob/main/packages/webview_flutter/webview_flutter_android/example/integration_test/webview_flutter_test.dart#L100

I don't think the implementation of the test should block this PR though. Leaks only prevent the release of the strong reference to the object in the Java `InstanceManager`. And the camera resources are explicitly released with the methods provided by the API.

I can also help with writing the test.

File: packages/camera/camera_android_camerax/test/observer_test.dart
Comment (reidbaker) [inline packages/camera/camera_android_camerax/test/observer_test.dart:113]: Can you add a test to make sure createFromInstances copies the values correctly? 

Same for other classes if a similar method exists.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart]: I made them both ! because neither should be null!

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:593]: This is handled in the `Observer` implementation. Updated it from an assert statement there to throwing an error, though

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart]: Makes perfect sense. Renamed and went with option B

File: packages/camera/camera_android_camerax/lib/src/camera_info.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/camera_info.dart]: Oops! I meant to remove all asserts. Removed it

File: packages/camera/camera_android_camerax/lib/src/live_data.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/live_data.dart:185]: Added a default case with an error. Needed an ignore statement to do that, but left a comment why it was safe

File: packages/camera/camera_android_camerax/test/observer_test.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/test/observer_test.dart:113]: This is handled in the host API create tests. createFromInstances calls through to the mockApis so when we test that call, we ensure createFromInstances is handling the values correctly

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:592]: Added to https://github.com/flutter/flutter/issues/125928

**Review by bparrishMines — APPROVED** [review]
LGTM

**Review by reidbaker — APPROVED** [review]

File: packages/camera/camera_android_camerax/lib/src/observer.dart
Comment (reidbaker) [reply packages/camera/camera_android_camerax/lib/src/observer.dart]: This was marked resolved but I dont seee anything explaining why it should be plural. Non blocking but I am still curious for the answer.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (reidbaker) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart]: Resolved but not a constant and no comment explaining why.

Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart]: I think you're looking at outdated code?

File: packages/camera/camera_android_camerax/lib/src/observer.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/observer.dart]: I think you're also looking at outdated code here; I changed it to singular.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:38]: Yeah, it's still possible to change from non-testing code :/

How would I change this from testing code, though, with this structure? I need a way to set this to true for testing, so is there a way to do that without exposing it also to non-testing code? I would think using a setter, but that's still visible.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:502]: First issue: @bparrishMines am I misunderstanding how weakThis should be used? Should I allow for the target to be null to avoid leaks?

Second issue: Yes, this is a helper method used to set up the image streaming that is called from listening to the stream returned by onStreamedFrameAvailable, so it should never be called really any other time. cameraImageDataStreamController is set up before that, so it would never be null here.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:575]: I think this gets at the bigger question as to why `cameraId` is used throughout the Flutter camera API to begin with. Really, `cameraId` is the ID of the Flutter surface texture used to render a preview to, and thus, is almost totally unrelated to the actual `camera` instance we use to do any camera functionality. The Flutter surface texture ID is returned by calls to `createCamera(...)`, so I assume it is just an ID that folks using the plugin can mostly use to filter through `CameraEvents` added to the various camera event streams (because they are constructed with an ID) because otherwise, at least in the Android camera implementations, it goes mostly unused.

I'm on the fence about renaming `cameraId`, because while it would be more useful to name it something more like `flutterSurfaceTextureId`, it's not named that in the actual camera platform interface, so it could cause more confusion for folks trying to contribute to the plugin.

However, I am on board for bettering documentation in the implementation around what it is (always referring to the Flutter surface texture ID/`createCamera`), when it is actually being used in a method, and if it is being used, exactly how. I'll file an issue before landing this for that.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (reidbaker) [inline packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:594]:
```suggestion
      // This cast is safe because the Observer implementation ensures
```

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:576]: https://github.com/flutter/flutter/issues/126270 filed for improving documentation

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (bparrishMines) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:503]: I think the point of the weak reference is to ensure the callback method doesn't prevent garbage collection of the instance that contains it. The `target` COULD be null, but the `!` is just a runtime check that can be used if you expect the `target` to always be available in the callback.

For example, if you expect the user to eventually ALWAYS call `CameraController.dispose` and you do some kind of cleanup in that method that stops image streaming callbacks. Then I think it is reasonable to expect this to always be nonnull. But a test would also be helpful for ensuring this and preventing a null pointer exception.

As a comparison, `weview_flutter` always uses `weakThis.target?...` because there is no `WebViewController.dispose`.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart
Comment (reidbaker) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:38]: Here is a patch file with it working you can apply. 
[protected-testing-object.patch](https://github.com/flutter/packages/files/11424583/protected-testing-object.patch)

**Pull request merged**
